File: 38.cpp
#include <iostream>

using namespace std;

/*思路：利用前序遍历，到访问到叶结点时，比较最大，返回到父节点时，减去1*/
void PreSearch(TreeNode* pRoot, int& result, int& tmp)
{
    if (pRoot == NULL)
    {
        if (tmp > result)
            result = tmp;
        return;
    }
    tmp++;
    PreSearch(pRoot->left, result, tmp);
    PreSearch(pRoot->right, result, tmp);
    tmp--;
}

int TreeDepth(TreeNode* pRoot)
{
    int result = 0;
    if (pRoot == NULL)
        return result;
    int tmp = 0;
    PreSearch(pRoot, result, tmp);
    return result;
}

File: 38.py
# -*- coding:utf-8 -*-
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
class Solution:
    def TreeDepth(self, pRoot):
        # write code here
        if not pRoot:
            return 0
        result = [0]
        tmp = [0]
        self.PreSearch(pRoot, tmp, result)
        return result[0]
        
    def PreSearch(self, Root, tmp, result):
        if not Root:
            if tmp[0] > result[0]:
                result[0] = tmp[0]
            return
        tmp[0] += 1
        self.PreSearch(Root.left, tmp, result)
        self.PreSearch(Root.right, tmp, result)
        tmp[0] -= 1